validator/client/pan_propose.go

```diff
@@ -0,0 +1,274 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	eth1Types "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/pkg/errors"
+	types "github.com/prysmaticlabs/eth2-types"
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	validatorpb "github.com/prysmaticlabs/prysm/proto/validator/accounts/v2"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/timeutils"
+	"github.com/prysmaticlabs/prysm/validator/pandora"
+	"golang.org/x/crypto/sha3"
+	"time"
+)
+
+var (
+	// errInvalidHeaderHash is returned if the header hash does not match with incoming header hash
+	errInvalidHeaderHash = errors.New("invalid header hash")
+	// errInvalidSlot is returned if the current slot does not match with incoming slot
+	errInvalidSlot = errors.New("invalid slot")
+	// errInvalidEpoch is returned if the epoch does not match with incoming epoch
+	errInvalidEpoch = errors.New("invalid epoch")
+	// errInvalidProposerIndex is returned if the proposer index does not match with incoming proposer index
+	errInvalidProposerIndex = errors.New("invalid proposer index")
+	// errInvalidTimestamp is returned if the timestamp of a block is higher than the current time
+	errInvalidTimestamp = errors.New("invalid timestamp")
+	// errNilHeader
+	errNilHeader = errors.New("pandora header is nil")
+	// errPanShardingInfoNotFound
+	errPanShardingInfoNotFound = errors.New("pandora sharding info not found in canonical head")
+	// errSubmitShardingSignatureFailed
+	errSubmitShardingSignatureFailed = errors.New("pandora sharding signature submission failed")
+)
+
+// processPandoraShardHeader method does the following tasks:
+// - Get pandora block header, header hash, extraData from remote pandora node
+// - Validate block header hash and extraData fields
+// - Signs header hash using a validator key
+// - Submit signature and header to pandora node
+func (v *validator) processPandoraShardHeader(
+	ctx context.Context,
+	beaconBlk *ethpb.BeaconBlock,
+	slot types.Slot,
+	epoch types.Epoch,
+	pubKey [48]byte,
+) error {
+
+	fmtKey := fmt.Sprintf("%#x", pubKey[:])
+	latestPandoraHash := eth1Types.EmptyRootHash
+	latestPandoraBlkNum := uint64(0)
+
+	// if pandoraShard is nil means there is no pandora blocks in pandora chain except block-0
+	if beaconBlk.Body != nil && beaconBlk.Body.PandoraShard != nil {
+		latestPandoraHash = common.BytesToHash(beaconBlk.Body.PandoraShard[0].Hash)
+		latestPandoraBlkNum = beaconBlk.Body.PandoraShard[0].BlockNumber
+	}
+
+	// Request for pandora chain header
+	header, headerHash, extraData, err := v.pandoraService.GetShardBlockHeader(ctx, latestPandoraHash, latestPandoraBlkNum+1)
```

Comment:
I think that you should restrict yourself to the `highestPandoraHash` not `latestPandoraHash`